src/ray/raylet/worker_pool.h

```diff
@@ -476,13 +479,22 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
     /// the process. The shim process PID is the same with worker process PID, except
     /// starting worker process in container.
     std::unordered_map<Process, StartingWorkerProcessInfo> starting_worker_processes;
-    /// A map for looking up the task by the pid of starting worker process.
-    std::unordered_map<Process, TaskWaitingForWorkerInfo> starting_workers_to_tasks;
-    /// A map for looking up the task with dynamic options by the pid of
-    /// starting worker process. Note that this is used for the dedicated worker
-    /// processes.
-    std::unordered_map<Process, TaskWaitingForWorkerInfo>
-        starting_dedicated_workers_to_tasks;
+    /// Reverse index from RuntimeEnvHash (which identifies a starting worker)
+    /// to a task that can only run on that dedicated worker. This is used for
+    /// actor creation tasks that require dynamic worker creation options,
+    /// while the worker is still starting.
+    absl::flat_hash_map<RuntimeEnvHash, TaskID> starting_dedicated_workers_to_tasks;
+    /// A map from runtime env hash to tasks that require that runtime env.
+    /// The runtime envs for tasks in this map have been successfully created,
+    /// and the tasks are now waiting for a worker to start. Note that this map
+    /// does not include tasks that have a dedicated worker assigned to them.
+    absl::flat_hash_map<RuntimeEnvHash, std::deque<TaskWaitingForWorkerInfo>>
+        waiting_tasks_by_env_hash;
+    /// A map from runtime env hash to workers that are starting in that
+    /// runtime env. Note that this map does not include dedicated workers that
+    /// have already been assigned to a task.
+    absl::flat_hash_map<RuntimeEnvHash, std::unordered_set<Process>>
```

Comment:
Could we report the {key, value} sizes of these maps in WorkerPool::DebugString()?

Reply:
```suggestion
    absl::flat_hash_map<RuntimeEnvHash, absl::flat_hash_set<Process>>
```